### PR TITLE
SA11-038 Improve Refactoring Rename

### PR DESCRIPTION
**Pull request by joaopsazevedo**
The existing rename funcitonality was replaced by Laltools' one.
An error message is return if it is detected that the rename operation will
cause problems.
Added one test for a name collision problem.
Output of rename tests with the changes required were sorted by line number.

### CODE REVIEW

File: source/ada/lsp-ada_handlers.adb [hunk 1]
```diff
@@ -2647,81 +2648,96 @@ package body LSP.Ada_Handlers is
       Document  : constant LSP.Ada_Documents.Document_Access :=
                    Get_Open_Document (Self, Value.textDocument.uri);
 
+      Refs      : Laltools.Refactor.Rename.Renamable_References;
+
       procedure Process_Context (C : Context_Access);
       --  Process the rename request for the given context, and add
-      --  the results to response.
+      --  the results to response. If a rename problem is found, return a
```
Comment (setton) [inline source/ada/lsp-ada_handlers.adb:2655]: (clarification): this says "if a problem is found, return a response with an error": at first I thought this should be transformed into a function that actually returns the error... I suggest to clarify this by explaining which mechanism is being used to return the error to the user.

File: source/ada/lsp-ada_handlers.adb [hunk 2]
```diff
@@ -2647,81 +2648,96 @@ package body LSP.Ada_Handlers is
...
+         Name_Node : Name := Laltools.Common.Get_Node_As_Name
            (C.Get_Node_At (Document, Position));
 
-         Definition : Defining_Name;
-         Imprecise  : Boolean;
-         Empty      : LSP.Messages.TextEdit_Vector;
-         Count      : Cancel_Countdown := 0;
+         Empty     : LSP.Messages.TextEdit_Vector;
+
+         --  Analysis unit of the reference that is being added to Response.
```
Comment (setton) [inline source/ada/lsp-ada_handlers.adb:2672]: (style): single-line comments shouldn't end with `.`

File: source/ada/lsp-ada_handlers.adb [hunk 3]
```diff
@@ -2647,81 +2648,96 @@ package body LSP.Ada_Handlers is
...
+         --  they contain a renamed word.
 
-         procedure Callback
-           (Node   : Libadalang.Analysis.Base_Id;
-            Kind   : Libadalang.Common.Ref_Result_Kind;
-            Cancel : in out Boolean);
+         procedure Process_Reference
+           (Unit_Filename : String;
+            Sloc_Range    : Langkit_Support.Slocs.Source_Location_Range);
+         --  Add a reference to Response if it has not been added yet.
```
Comment (setton) [inline source/ada/lsp-ada_handlers.adb:2684]: (see above)

File: source/ada/lsp-ada_handlers.adb [hunk 4]
```diff
@@ -2647,81 +2648,96 @@ package body LSP.Ada_Handlers is
...
+            Sloc_Range    : Langkit_Support.Slocs.Source_Location_Range)
          is
+            use type VSS.Unicode.UTF16_Code_Unit_Count;
+
             Location : constant LSP.Messages.Location :=
-              Get_Node_Location (Node.As_Ada_Node);
+              (uri     => LSP.Ada_Contexts.File_To_URI
+                 (To_Unbounded_Wide_String
+                      (To_Wide_String (Unit_Filename))),
+               span    => LSP.Messages.Span'
```
Comment (setton) [inline source/ada/lsp-ada_handlers.adb:2700]: how about introducing an utility function to transform a Langkit Sloc_Range to a LSP span?

File: source/ada/lsp-ada_handlers.adb [hunk 5]
```diff
@@ -2647,81 +2648,96 @@ package body LSP.Ada_Handlers is
...
+               alsKind => <>);
             Item     : constant LSP.Messages.TextEdit :=
               (span    => Location.span,
                newText => Value.newName);
          begin
             if not Response.result.changes.Contains (Location.uri) then
                --  We haven't touched this document yet, create an empty
                --  change list
-               Response.result.changes.Insert (Location.uri, Empty);
+               Response.result.changes.Insert
```
Comment (setton) [inline source/ada/lsp-ada_handlers.adb:2717]: (style): no need for a line break here

File: source/ada/lsp-ada_handlers.adb [hunk 6]
```diff
@@ -2874,45 +2890,80 @@ package body LSP.Ada_Handlers is
...
-         if Name_Node = No_Name then
-            return;
-         end if;
+         use Laltools.Refactor.Rename;
 
-         Definition := Laltools.Common.Resolve_Name
-           (Name_Node,
-            Self.Trace,
-            Imprecise => Imprecise);
+         Cursor_1 : Unit_Slocs_Maps.Cursor;
```
Comment (setton) [inline source/ada/lsp-ada_handlers.adb:2895]: (style): name the cursors with something more meaningful, for instance `Unit_Cursor` and `Sloc_Cursor` ?

File: source/ada/lsp-lal_utils.adb [hunk 7]
```diff
@@ -545,6 +545,19 @@ package body LSP.Lal_Utils is
...
+   -- To_Unbounded_Text_Type --
+   ----------------------------
+
+   function To_Unbounded_Text_Type
+     (Item : LSP_String)
+      return Langkit_Support.Text.Unbounded_Text_Type
+   is
+      use Langkit_Support.Text;
+   begin
+      return To_Unbounded_Text (From_UTF8 (To_UTF_8_String (Item)));
```
Comment (setton) [inline source/ada/lsp-lal_utils.adb:558]: It's dangerous to return a String when the string might be giant: add a warning to say we shouldn't pass a buffer to this function.

**Review by setton — CHANGES_REQUESTED** [review]

**Review by AnthonyLeonardoGracio — APPROVED** [review]

**Review by setton — APPROVED** [review]

**Pull request merged**